### DISCUSSION TITLE
"Hello, is there a method in iTwin.js for directly transforming a model into the current world coordinates?"

### DISCUSSION THREAD

**Discussion by MYDJH** [iTwin/itwinjs-core#7188] [Q&A] [answered]
"I read the geometric information of this object using the following code."
```js
 const it = new GeometryStreamIterator(props.geom);
        for (const entry of it) {
          console.log(
            entry.primitive.geometry,
            index,
            "entry"
          );
Let curve =  entry.primitive.geometry
}
//Add the model to the view.
      decoratorState.setLinePixels(LinePixels.Solid);
     decoratorState.addGeometry(curve);
```
"However, when I add this entry.primitive.geometry to the model, I find that the position of the added model does not match the position of the original model."

"The position of the added model and the actual position of the model."
![image](https://github.com/user-attachments/assets/eef88604-3a04-4dbd-ba1c-6a5115f0d4a1)

the geometric information read from entry.primitive.geometry
![image](https://github.com/user-attachments/assets/7dc9ce8f-648d-4b03-9c42-b2d0334c3609)

"I found that the actual position of the model needs to be adjusted by adding the element's origin to get the correct position. Now, I want to directly modify the position of the entry.primitive.geometry object. Does iTwin.js provide such a coordinate transformation?"

 the element's origin:
![image](https://github.com/user-attachments/assets/73892b9a-95ae-473e-bd0d-b507a88bd727)

"The actual model coordinates need to be adjusted by adding x + element.origin.x, y + element.origin.y, and z + element.origin.z. How can I solve this problem? Thank you very much for reading, and I would greatly appreciate any good suggestions you may have."

**Comment by bbastings** [comment, marked as answer]
The geometry stream entries are stored relative to the element's placement. The easiest way to iterate over the geometry in world coordinates is to create the iterator from a GeometricElementProps. [World Coordinate Iterator from GeometricElement3d](https://www.itwinjs.org/reference/core-common/geometry/geometrystreamiterator/fromgeometricelement3dstatic/)

Similarly, if you were to create a new element, it's geometry should also be stored relative to a meaningful placement and not world coordinates. GeometryStreamBuilder can automatically transform world coordinate geometry relative to the desired placement using one of the method to specify the local-to-world transform [such as](https://www.itwinjs.org/reference/core-common/geometry/geometrystreambuilder/setlocaltoworldfromplacement/). If you were creating a new element from the geometry extracted from an existing element, it might make sense to create the new element using the same placement as the existing element, etc.  _You aren't though, looks like you are just creating a view decorations._

More information about creating geometric element's and placement [here](https://www.itwinjs.org/learning/common/geometrystream/).

**Reply by MYDJH** [reply]
Thank you very much.
